### PR TITLE
feat: add wechat mini program support

### PR DESCRIPTION
**Pull request by Steve0x2a**
Fix: https://github.com/casdoor/casdoor/issues/400
Thanks to @mojerro 's help, I was able to develop the WeChat mini program without being familiar with it.
This PR currently only implements the most simple WeChat mini program access function, if anyone needs more specific features, please feel free to comment.
First you have to create a wechat idp in casdoor:
image

then login in mini program
```js
// login in mini program
    // 登录
    wx.login({
      success: res => {
        // 发送 res.code 到后台换取 openId, sessionKey, unionId
        console.log(res.code)
        
        wx.request({
          url: 'http://localhost:8000/api/login/oauth/access_token',
          method: "POST",
          data: {
            "tag":"wechat_miniprogram", //required
            "client_id":"6825f4f0af45554c8952",
            "client_secret": "2d0f463...",
            "code": res.code,
            "username": this.data.userInfo.nickName, //update user profile, when you login.
            "avatar": this.data.userInfo.avatarUrl,
          },
          header:{
            "content-type":"application/x-www-form-urlencoded",
          },
          success: res => {
            console.log(res)
            this.globalData.accessToken = res.data.access_token
          }
        })
      }
    })
```
now you can use accessToken with `app.globalData.accessToken`
For example, updating your user information to Casdoor:
```js
    wx.getUserProfile({
      desc: '展示用户信息', // 声明获取用户个人信息后的用途，后续会展示在弹窗中，请谨慎填写
      success: (res) => {
        this.setData({
          userInfo: res.userInfo,
          hasUserInfo: true
        })
        console.log(app.globalData.accessToken)
        wx.request({
          url: 'http://localhost:7001/api/update-user',
          method: "POST",
          data: {
            "owner": "test",
            "name": "wechat-oGk3T5tIiMFo3SazCO75f0HEiE7Q",
            "displayName": this.data.userInfo.nickName,
            "avatar": this.data.userInfo.avatarUrl
          },
          header:{
            "Authorization":"Bearer "+app.globalData.accessToken, //Bearer token
            "content-type": "application/json"
          },
          success: (res)=>{
            console.log(res)
          }
        })
      }
    })
```

Signed-off-by: Steve0x2a <stevesough@gmail.com>

### CODE REVIEW

File: object/token.go [hunk 1]
```diff
@@ -608,3 +611,59 @@ func GetTokenByUser(application *Application, user *User, scope string, host str
...
+		}
+		AddUser(user)
+	}
+
+	accessToken, refreshToken, err := generateJwtToken(application, user, "", "", host)
+	if err != nil {
+		return nil, err
+	}
+
+	token := &Token{
```
Comment (mojerro) [inline object/token.go:652]: I think the JWT token can add a ClientID field, which storage  the openId or unionId

File: object/token.go [hunk 2]
```diff
@@ -608,3 +611,59 @@ func GetTokenByUser(application *Application, user *User, scope string, host str
...
+		Organization: user.Owner,
+		User:         user.Name,
+		Code:         session.Session_key, //a trick, because miniprogram does not use the code, so use the code field to save the session_key
+		AccessToken:  accessToken,
+		RefreshToken: refreshToken,
+		ExpiresIn:    application.ExpireInHours * 60,
+		Scope:        "",
+		TokenType:    "Bearer",
+		CodeIsUsed:   true,
+	}
```
Comment (mojerro) [inline object/token.go:666]: I think the JWT token can add a ClientID field, which storage  the openId or unionId.

just like the wechat token now. 

`wechat.go`
```go
type WechatAccessToken struct {
	AccessToken  string `json:"access_token"`  //Interface call credentials
	ExpiresIn    int64  `json:"expires_in"`    //access_token interface call credential timeout time, unit (seconds)
	RefreshToken string `json:"refresh_token"` //User refresh access_token
	Openid       string `json:"openid"`        //Unique ID of authorized user
	Scope        string `json:"scope"`         //The scope of user authorization, separated by commas. (,)
	Unionid      string `json:"unionid"`       //This field will appear if and only if the website application has been authorized by the user's UserInfo.
}
```

File: object/token.go [hunk 3]
```diff
@@ -608,3 +611,59 @@ func GetTokenByUser(application *Application, user *User, scope string, host str
...
+			CreatedTime:       util.GetCurrentTime(),
+			IsAdmin:           false,
+			IsGlobalAdmin:     false,
+			IsForbidden:       false,
+			IsDeleted:         false,
+		}
+		AddUser(user)
+	}
+
+	accessToken, refreshToken, err := generateJwtToken(application, user, "", "", host)
```
Comment (mojerro) [inline object/token.go:647]: I'm not familiar with the code and hope to confirm that is the JWT token has openId and unionId fields same as the wechat file implemented now.

```go
// idp/wechat.go
type WechatAccessToken struct {
	AccessToken  string `json:"access_token"`  //Interface call credentials
	ExpiresIn    int64  `json:"expires_in"`    //access_token interface call credential timeout time, unit (seconds)
	RefreshToken string `json:"refresh_token"` //User refresh access_token
	Openid       string `json:"openid"`        //Unique ID of authorized user
	Scope        string `json:"scope"`         //The scope of user authorization, separated by commas. (,)
	Unionid      string `json:"unionid"`       //This field will appear if and only if the website application has been authorized by the user's UserInfo.
}
```

Comment (Steve0x2a) [reply object/token.go:647]: @leopku in idp/wechat.go, if the return has a unionID, then use the unionID, if only the openID, then use the openID, so I have the following code in this pr:
```go
existed, err := adapter.Engine.Where("wechat = ? OR wechat = ? OR unionid = ?", openId, unionId, unionId).Get(user)
```

Comment (mojerro) [reply object/token.go:647]: thanks for your explanation

File: object/user.go [hunk 4]
```diff
@@ -76,6 +76,7 @@ type User struct {
 	Google   string `xorm:"varchar(100)" json:"google"`
 	QQ       string `xorm:"qq varchar(100)" json:"qq"`
 	WeChat   string `xorm:"wechat varchar(100)" json:"wechat"`
+	Unionid  string `xorm:"varchar(100)" json:"unionid"`
```
Comment (hsluoyz) [inline object/user.go:79]: WeChatUnionId

File: object/user.go [hunk 5]
```diff
@@ -227,6 +228,23 @@ func getUserById(owner string, id string) *User {
 	}
 }
 
+func getUserByOpenId(openId, unionId string) *User {
```
Comment (hsluoyz) [inline object/user.go:231]: Don't omit string

File: idp/wechat_miniprogram.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"golang.org/x/oauth2"
+)
+
+type WeChatMPIdProvider struct {
```
Comment (hsluoyz) [inline idp/wechat_miniprogram.go:26]: WeChatMPIdProvider -> WeChatMiniProgramIdProvider

File: idp/wechat_miniprogram.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	idp.Client = &http.Client{}
+	return idp
+}
+
+func (idp *WeChatMPIdProvider) SetHttpClient(client *http.Client) {
+	idp.Client = client
+}
+
+func (idp *WeChatMPIdProvider) getConfig(clientId string, clientSecret string) *oauth2.Config {
+
```
Comment (hsluoyz) [inline idp/wechat_miniprogram.go:45]: no line

File: idp/wechat_miniprogram.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	var config = &oauth2.Config{
+		ClientID:     clientId,
+		ClientSecret: clientSecret,
+	}
+
+	return config
+}
+
+type WeChatMPSeesionResponse struct {
```
Comment (hsluoyz) [inline idp/wechat_miniprogram.go:54]: WeChatMPIdProvider -> WeChatMiniProgramIdProvider

File: idp/wechat_miniprogram.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+	Openid      string `json:"openid"`
+	Session_key string `json:"session_key"`
```
Comment (hsluoyz) [inline idp/wechat_miniprogram.go:56]: SessionKey

File: idp/wechat_miniprogram.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	Unionid     string `json:"unionid"`
+	Errcode     int    `json:"errcode"`
+	Errmsg      string `json:"errmsg"`
+}
+
+func (idp *WeChatMPIdProvider) GetSeesionByCode(code string) (*WeChatMPSeesionResponse, error) {
```
Comment (hsluoyz) [inline idp/wechat_miniprogram.go:62]: All `Seesion` all typos.

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/ApplicationEditPage.js [hunk 11]
```diff
@@ -455,6 +455,7 @@ class ApplicationEditPage extends React.Component {
                           {id: "client_credentials", name: "Client Credentials"},
                           {id: "token", name: "Token"},
                           {id: "id_token",name:"ID Token"},
+                          {id: "wechat_miniprogram", name:"WeChat Mini Program"}
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:458]: Why add here?

File: object/user.go (hunk 5 above)
Comment (hsluoyz) [reply object/user.go:231]: Never use `OpenId` only, add WeChat prefix

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

**Comment by hsluoyz** [comment]
@mojerro plz review

**Pull request merged**